packages/flutter/lib/src/material/navigation_bar.dart

```diff
@@ -0,0 +1,1144 @@
+      // direction.
+      child: StatusTransitionWidgetBuilder(
+        animation: animation,
+        builder: (BuildContext context, Widget? child) {
+          return SelectableAnimatedBuilder(
+            isSelected: animation.isForwardOrCompleted,
+            duration: const Duration(milliseconds: 100),
+            alwaysDoFullAnimation: true,
+            builder: (BuildContext context, Animation<double> fadeAnimation) {
+              return FadeTransition(
+                opacity: fadeAnimation,
+                child: Container(
+                  width: 64,
+                  height: 32,
+                  decoration: BoxDecoration(
+                    borderRadius: BorderRadius.circular(20),
+                    color: color ?? colorScheme.secondary.withOpacity(.24),
+                  ),
+                ),
+              );
+            },
+          );
+        },
+      ),
+    );
+  }
+}
+
+/// Inherited widget for passing data from the [NavigationBar] to the
+/// [NavigationBar.destinations] children widgets.
+///
+/// Useful for building navigation bar destinations using:
+/// `NavigationBarDestinationInfo.of(context)`.
+class NavigationBarDestinationInfo extends InheritedWidget {
+  /// Adds the information needed to build a navigation bar destination to the
+  /// [child] and descendants.
+  const NavigationBarDestinationInfo({
+    Key? key,
+    required this.destinationNumber,
+    required this.totalNumberOfDestinations,
+    required this.selectedAnimation,
+    required this.labelBehavior,
+    required this.onTap,
+    required Widget child,
+  }) : super(key: key, child: child);
+
+  /// Which destination number is this in the navigation bar.
+  ///
+  /// For example:
+  /// ```dart
+  /// NavigationBar(
+  ///   destinations: [
+  ///     NavigationBarDestination(), // This is destination number 0.
+  ///     NavigationBarDestination(), // This is destination number 1.
+  ///     NavigationBarDestination(), // This is destination number 2.
+  ///   ]
+  /// )
+  /// ```
+  ///
+  /// This is required for semantics, so that each destination can have a label
+  /// "Tab 1 of 3", for example.
+  final int destinationNumber;
+
+  /// How many total destinations are are in this navigation bar.
+  ///
+  /// This is required for semantics, so that each destination can have a label
+  /// "Tab 1 of 4", for example.
+  final int totalNumberOfDestinations;
+
+  /// Indicates whether or not this destination is selected, from 0 (unselected)
+  /// to 1 (selected).
+  final Animation<double> selectedAnimation;
+
+  /// Determines the behavior for how the labels will layout.
+  ///
+  /// Can be used to show all labels (the default), show only the selected
+  /// label, or hide all labels.
+  final NavigationBarDestinationLabelBehavior labelBehavior;
+
+  /// The callback that should be called when this destination is tapped.
+  ///
+  /// This is computed by calling [NavigationBar.onTap] with [destinationNumber]
```

Comment:
NavigationBar.onDestinationSelected, should probably be consistent with what we call the handler